### PR TITLE
PCHR-2749: Close quick search field on outside click + refactoring

### PR DESCRIPTION
**Pull request by AkA84**
## Overview
This PR improves the behaviour of the collapsible quick search field,

## Before
The field does not get collapsed when the user clicks outside of it

before

## After
Now the field gets collapsed with an outside click

after

## Comments
This happens only if the field is empty. Otherwise the field remains visible (as it's the expected behaviour)

### CODE REVIEW

File: hrui/js/src/hrui.js [hunk 1]
```diff
@@ -140,8 +140,23 @@
...
-    toggleCustomClassToQuickSearchField();
     changeQuickSearchFieldPlaceholder();
+    manageCustomClassOfQuickSearchField();
+  }
+
+  /**
+   * Checks if the user has clicked outside of the quick search field
+   * by analyzing the given click target
+   *
+   * @param {object} target
```
Comment (igorpavlov) [inline hrui/js/src/hrui.js]: Could it be better to state `@param {Element} target` instead?

**Comment by igorpavlov** [comment]
@AkA84 Any particular reason why not smth like that?:
```js
$elem.blur(function () {
   if ($elem.val() === '') $elem.hide();
});
```

File: hrui/js/src/hrui.js [hunk 2]
```diff
@@ -140,8 +140,23 @@
...
+   * @return {boolean}
+   */
+  function hasUserClickedOutsideQuickSearchField (target) {
+    $target = $(target);
```
Comment (igorpavlov) [inline hrui/js/src/hrui.js]: Should be `var $target = ...`

I wonder why semistandard didn't yell? Or you can mutate `target = ...` to save some memory.

File: hrui/js/src/hrui.js [hunk 3]
```diff
@@ -140,8 +140,23 @@
...
+
+    return !$target.is('#crm-qsearch') &&
+      !$target.is('#root-menu-div') &&
+      !$target.parents('#crm-qsearch, #root-menu-div').length;
```
Comment (igorpavlov) [inline hrui/js/src/hrui.js]: What about `!$target.closest('#crm-qsearch, #root-menu-div').length;` ?

File: hrui/js/src/hrui.js [hunk 4]
```diff
@@ -160,6 +175,15 @@
   }
 
   /**
+   * Checks if the quick search field itself has any current value
```
Comment (igorpavlov) [inline hrui/js/src/hrui.js]: Checks if the quick search field currently has any value

File: hrui/js/src/hrui.js [hunk 5]
```diff
@@ -160,6 +175,15 @@
...
+   *
+   * @return {boolean}
+   */
+  function isQuickSearchOnGoing () {
+    return $('#sort_name_navigation').val().length;
```
Comment (igorpavlov) [inline hrui/js/src/hrui.js]: what if `$('#sort_name_navigation').val() === '        '` don't we need to trim()?

this is not a change request, up to you.

File: hrui/js/src/hrui.js [hunk 6]
```diff
@@ -197,6 +232,22 @@
   }
 
   /**
+   * Removes the given custom class when the user clicks
+   * outside the quick search field (if there is no ongoing search)
+   *
+   * @param {string} customClass
+   */
+  function removeCustomClassOnOutsideClick (customClass) {
+    $('body').on('click', function (event) {
```
Comment (igorpavlov) [inline hrui/js/src/hrui.js]: $(document) would be safer in case `body` has margins.

File: hrui/js/src/hrui.js [hunk 7]
```diff
@@ -197,6 +232,22 @@
...
+      $target = $(event.target);
```
Comment (igorpavlov) [inline hrui/js/src/hrui.js]: Same as above. It should yell $target is not defined.

File: hrui/js/src/hrui.js (hunk 2 above)
Comment (AkA84) [reply hrui/js/src/hrui.js]: wow, that's bad. I also forgot to enable the git hook in my local env, will do now

File: hrui/js/src/hrui.js (hunk 5 above)
Comment (AkA84) [reply hrui/js/src/hrui.js]: changed it to
```js
return !!$('#sort_name_navigation').val().trim();
```

**Pull request merged**